File: bindings/kotlin/lib/src/main/kotlin/org/equalitie/ouisync/Repository.kt
package org.equalitie.ouisync

import org.msgpack.core.MessagePacker

/**
 *
 * A Ouisync repository.
 *
 * Example usage:
 *
 * ```
 * // Create a new repo:
 * val repo = Repository.create(session, "path/to/the/repo.ouisyncdb")
 *
 * // or open an existing one:
 * val repo = Repository.open(session, "path/to/the/repo.ouisyncdb")
 *
 * // Access the repository files (see File, Directory) ...
 *
 * // Close it when done:
 * repo.close()
 * ```
 *
 * To share a repository, create the share token with [createShareToken], send it to the peer
 * (e.g., via a secure instant messenger, encode as QR code and scan, ...), then create a
 * repository on the peer's device with [create], passing the share token to it.
 *
 * @see Session
 * @see File
 * @see Directory
 */
class Repository private constructor(internal val handle: Long, internal val client: Client) {
    companion object {
        /**
         * Creates a new repository.
         *
         * @param session       the Ouisync session.
         * @param path          path to the local file to store the repository in. It's recommended
         *                      to use the "ouisyncdb" file extension, but any extension works.
         * @param readPassword  local password for reading the repository on this device only. Do
         *                      not share with peers!. If null, the repo won't be password
         *                      protected and anyone with physical access to the device will be
         *                      able to read it.
         * @param writePassword local password for writing to the repository on this device only.
         *                      Do not share with peers! Can be the same as `readPassword` if one
         *                      wants to use only one password for both reading and writing.
         *                      Separate passwords are useful for plausible deniability. If both
         *                      `readPassword` and `writePassword` are null, the repo won't be
         *                      password protected and anyone with physical access to the device
         *                      will be able to read and write to it. If `readPassword` is not null
         *                      but `writePassword` is null, the repo won't be writable from this
         *                      device.
         * @param shareToken    used to share repositories between devices. If not null, this repo
         *                      will be linked with the repos with the same share token on other
         *                      devices. See also [createShareToken]. This also determines the
         *                      maximal access mode the repo can be opened in. If null, it's *write*
         *                      mode.
         */
        suspend fun create(
            session: Session,
            path: String,
            readPassword: String?,
            writePassword: String?,
            shareToken: ShareToken? = null,
        ): Repository {
            val client = session.client
            val handle = client.invoke(
                RepositoryCreate(
                    path,
                    readPassword,
                    writePassword,
                    shareToken?.toString(),
                ),
            ) as Long

            return Repository(handle, client)
        }

        /**
         * Open an existing repository.
         *
         * @param session  the Ouisync session.
         * @param path     path to the local file the repo is stored in.
         * @param password a local password. See the `readPassword` and `writePassword` param in
         *                 [create] for more details. If this repo uses local password(s), this
         *                 determines the access mode the repo is opened in: `readPassword` opens it
         *                 in *read* mode, `writePassword` opens it in *write* mode and no password
         *                 or wrong password opens it in *blind* mode. If this repo doesn't use
         *                 local password(s), the repo is opened in the maximal mode specified when
         *                 the repo was created.
         */
        suspend fun open(session: Session, path: String, password: String? = null): Repository {
            val client = session.client
            val handle = client.invoke(RepositoryOpen(path, password)) as Long

            return Repository(handle, client)
        }
    }

    /**
     * Closes the repository.
     */
    suspend fun close() = client.invoke(RepositoryClose(handle))

    /**
     * Subscribe to repository events.
     *
     * An even is emitted every time the content of the repo changes (e.g., a file is created,
     * written to, removed, moved, ...).
     */
    suspend fun subscribe(): EventReceiver<Unit> =
        client.subscribe(RepositorySubscribe(handle))

    /**
     * Returns the info-hash used to announce the repo on the Bittorrent DHT.
     *
     * Note the announce happens automatically. This function exists just for information/debugging
     * purposes.
     */
    suspend fun infoHash() = client.invoke(RepositoryInfoHash(handle)) as String

    /**
     * Returns the *database id* of this repository. A *database id* remains unchanged even when
     * the repo is renamed or moved and so can be used e.g. as a key for storing any per-repo
     * configuration, if needed.
     */
    suspend fun databaseId() = client.invoke(RepositoryDatabaseId(handle)) as ByteArray

    /**
     * Creates a *share token* to share this repository with other devices.
     *
     * By default the access mode of the token will be the same as the mode the repo is currently
     * opened in but it can be escalated with the `password` param or de-escalated with the
     * `accessMode` param.
     *
     * @param password   local repo password. If not null, the share token's access mode will be
     *                   the same as what the password provides. Useful to escalate the access mode
     *                   to above of what the repo is opened in.
     * @param accessMode access mode of the token. Useful to de-escalate the access mode to below
     *                   of what the repo is opened in.
     * @param name       optional human-readable name of the repo that the share token will be
     *                   labeled with. Useful to help organize the share tokens.
     */
    suspend fun createShareToken(
        password: String? = null,
        accessMode: AccessMode = AccessMode.WRITE,
        name: String? = null,
    ): ShareToken {
        val raw = client.invoke(RepositoryCreateShareToken(handle, password, accessMode, name)) as String
        return ShareToken(raw, client)
    }

    /**
     * Is local password required to read this repo?
     */
    suspend fun requiresLocalPasswordForReading() =
        client.invoke(RepositoryRequiresLocalPasswordForReading(handle)) as Boolean

    /**
     * Is local password required to write to this repo?
     */
    suspend fun requiresLocalPasswordForWriting() =
        client.invoke(RepositoryRequiresLocalPasswordForWriting(handle)) as Boolean

    /**
     * Returns the access mode (*blind*, *read* or *write*) the repo is opened in.
     */
    suspend fun accessMode(): AccessMode {
        val raw = client.invoke(RepositoryAccessMode(handle)) as Byte
        return AccessMode.decode(raw)
    }

    /**
     * Switches the repository to the given access mode.
     */
    suspend fun setAccessMode(
        accessMode: AccessMode,
        password: String?,
    ) =
        client.invoke(RepositorySetAccessMode(handle, accessMode, password))

    /**
     * Gets the current credentials of this repository. Can be used to restore access after
     * closing and reopening the repository.
     *
     * @see setCredentials
     */
    suspend fun credentials(): ByteArray = client.invoke(RepositoryCredentials(handle)) as ByteArray

    /**
     * Sets the current credentials of the repository.
     */
    suspend fun setCredentials(credentials: ByteArray) = client.invoke(RepositorySetCredentials(handle, credentials))

    /**
     * Sets, unsets or changes local passwords for accessing the repository or disables the given
     * access mode.
     */
    suspend fun setAccess(read: AccessChange? = null, write: AccessChange? = null) =
        client.invoke(RepositorySetAccess(handle, read, write))

    /**
     * Is Bittorrent DHT enabled?
     *
     * @see setDhtEnabled
     */
    suspend fun isDhtEnabled() = client.invoke(RepositoryIsDhtEnabled(handle)) as Boolean

    /**
     * Enables/disabled Bittorrent DHT (for peer discovery).
     *
     * @see isDhtEnabled
     * @see infoHash
     */
    suspend fun setDhtEnabled(enabled: Boolean) =
        client.invoke(RepositorySetDhtEnabled(handle, enabled))

    /**
     * Is Peer Exchange enabled?
     *
     * @see setPerEnabled
     */
    suspend fun isPexEnabled() = client.invoke(RepositoryIsPexEnabled(handle)) as Boolean

    /**
     * Enables/disables Peer Exchange (for peer discovery).
     *
     * @see isPexEnabled
     */
    suspend fun setPexEnabled(enabled: Boolean) =
        client.invoke(RepositorySetPexEnabled(handle, enabled))

    /**
     * Returns the synchronization progress of this repository
     */
    suspend fun syncProgress() = client.invoke(RepositorySyncProgress(handle)) as Progress

    /**
     * Create mirror of this repository on the cache server(s) that were previously added with
     * [Session.addCacheServer].
     */
    suspend fun mirror() = client.invoke(RepositoryMirror(handle))

    /**
     * Returns the type (file or directory) of an entry at the given path,
     * or `null` if no such entry exists.
     *
     * @param path path of the entry, relative to the repo root.
     */
    suspend fun entryType(path: String): EntryType? {
        val raw = client.invoke(RepositoryEntryType(handle, path)) as Byte?

        if (raw != null) {
            return EntryType.decode(raw)
        } else {
            return null
        }
    }

    /**
     * Moves an entry (file or directory) from `src` to `dst`.
     *
     * @param src path to move the entry from.
     * @param dst path to move the entry to.
     */
    suspend fun moveEntry(src: String, dst: String) =
        client.invoke(RepositoryMoveEntry(handle, src, dst))
}

/**
 * How to change access to a repository.
 *
 * @see [Repository.setAccess]
 */
sealed class AccessChange {
    fun pack(packer: MessagePacker) {
        when (this) {
            is EnableAccess -> {
                packer.packMapHeader(1)
                packer.packString("enable")

                if (password != null) {
                    packer.packString(password)
                } else {
                    packer.packNil()
                }
            }
            is DiableAccess -> {
                packer.packString("disabled")
            }
        }
    }
}

/**
 * Enable read or write access, optionally with local password
 *
 * @see [Repository.setAccess]
 */
class EnableAccess(val password: String?) : AccessChange()

/**
 * Disable access
 *
 * @see [Repository.setAccess]
 */
class DiableAccess() : AccessChange()

File: bindings/swift/OuisyncLib/Sources/Repository.swift
import Foundation
import MessagePack


public extension Client {
     /** Creates a new repository (or imports an existing repository) with optional local encryption
      *
      * If a `token` is provided, the operation will be an `import`. Otherwise, a new, empty,
      * fully writable repository is created at `path`.
      *
      * The optional `readSecret` and `writeSecret` are intended to function as a second
      * authentication factor and are used to encrypt the repository's true access keys. Secrets
      * are ignored when the underlying `token` doesn't contain the corresponding key (e.g. for
      * "blind" or "read-only" tokens).
      *
      * Finally, due to our current key distribution mechanism, `writeSecret` becomes mandatory
      * when a `readSecret` is set. You may reuse the same secret for both values. */
    func createRepository(at path: String,
                          importingFrom token: ShareToken? = nil,
                          readSecret: Secret? = nil,
                          writeSecret: Secret? = nil) async throws {
        // FIXME: the backend does buggy things here, so we bail out; see also `unsafeSetSecrets`
        if readSecret != nil && writeSecret == nil { throw OuisyncError.Unsupported }
        try await invoke("repository_create", with: ["path": .string(path),
                                                     "read_secret": readSecret?.value ?? .nil,
                                                     "write_secret": writeSecret?.value ?? .nil,
                                                     "token": token?.value ?? .nil,
                                                     "sync_enabled": false,
                                                     "dht_enabled": false,
                                                     "pex_enabled": false])
    }

    /** Opens an existing repository from a `path`, optionally using a known `secret`.
     *
     * If the same repository is opened again, a new handle pointing to the same underlying
     * repository is returned. Closed automatically when all references go out of scope. */
    func openRepository(at path: String, using secret: Secret? = nil) async throws -> Repository {
        try await Repository(self, invoke("repository_open", with: ["path": .string(path),
                                                                    "secret": secret?.value ?? .nil,
                                                                    "sync_enabled": false]))
    }

    /// All currently open repositories.
    var repositories: [Repository] { get async throws {
        try await invoke("repository_list").arrayValue.orThrow.map { try Repository(self, $0) }
    } }
}


public class Repository {
    let client: Client
    let handle: MessagePackValue
    init(_ client: Client, _ handle: MessagePackValue) throws {
        _ = try handle.uintValue.orThrow
        self.client = client
        self.handle = handle
    }

    deinit {
        // we're going out of scope so we need to copy the state that the async closure will capture
        let client = client, handle = handle
        Task { try await client.invoke("repository_close", with: handle) }
    }
}


public extension Repository {
    /// Deletes this repository. It's an error to invoke any operation on it after it's been deleted.
    func delete() async throws {
        try await client.invoke("repository_delete", with: handle)
    }

    func move(to location: String) async throws {
        try await client.invoke("repository_move", with: ["repository": handle,
                                                          "to": .string(location)])
    }

    var path: String { get async throws {
        try await client.invoke("repository_get_path", with: handle).stringValue.orThrow
    } }

    /// Whether syncing with other replicas is enabled.
    var syncing: Bool { get async throws {
        try await client.invoke("repository_is_sync_enabled", with: handle).boolValue.orThrow
    } }

    /// Enables or disables syncing with other replicas.
    func setSyncing(enabled: Bool) async throws {
        try await client.invoke("repository_set_sync_enabled", with: ["repository": handle,
                                                                      "enabled": .bool(enabled)])
    }

    /** Resets access using `token` and reset any values encrypted with local secrets to random
     * values. Currently that is only the writer ID. */
    func resetAccess(using token: ShareToken) async throws {
        try await client.invoke("repository_reset_access", with: ["repository": handle,
                                                                  "token": token.value])
    }

    /** The current repository credentials.
     *
     * They can be used to restore repository access via `setCredentials()` after the repo has been
     * closed and re-opened, without needing the local secret (e.g. when moving the database). */
    var credentials: Data { get async throws {
        try await client.invoke("repository_credentials", with: handle).dataValue.orThrow
    } }
    func setCredentials(from credentials: Data) async throws {
        try await client.invoke("repository_set_credentials",
                                with: ["repository": handle, "credentials": .binary(credentials)])
    }

    var accessMode: AccessMode { get async throws {
        try await AccessMode(rawValue: client.invoke("repository_get_access_mode",
                                                     with: handle).uint8Value.orThrow).orThrow
    } }
    func setAccessMode(to mode: AccessMode, using secret: Secret? = nil) async throws {
        try await client.invoke("repository_set_access_mode",
                                with: ["repository": handle,
                                       "access_mode": .uint(UInt64(mode.rawValue)),
                                       "secret": secret?.value ?? .nil])
    }

    /// Returns the `EntryType` at `path`, or `nil` if there's nothing that location
    func entryType(at path: String) async throws -> EntryType? {
        let res = try await client.invoke("repository_entry_type", with: ["repository": handle,
                                                                          "path": .string(path)])
        if case .nil = res { return nil } // FIXME: this should just be an enum type
        return try EntryType(rawValue: res.uint8Value.orThrow).orThrow
    }

    /// Returns whether the entry (file or directory) at `path` exists.
    @available(*, deprecated, message: "use `entryType(at:)` instead")
    func entryExists(at path: String) async throws -> Bool { try await entryType(at: path) != nil }

    /// Move or rename the entry at `src` to `dst`
    func moveEntry(from src: String, to dst: String) async throws {
        try await client.invoke("repository_move_entry", with: ["repository": handle,
                                                                "src": .string(src),
                                                                "dst": .string(dst)])
    }

    /// This is a lot of overhead for a glorified event handler
    @MainActor var events: AsyncThrowingMapSequence<Client.Subscription, Void> {
        client.subscribe(to: "repository").map { _ in () }
    }

    var dht: Bool { get async throws {
        try await client.invoke("repository_is_dht_enabled", with: handle).boolValue.orThrow
    } }
    func setDht(enabled: Bool) async throws {
        try await client.invoke("repository_set_dht_enabled", with: ["repository": handle,
                                                                     "enabled": .bool(enabled)])
    }

    var pex: Bool { get async throws {
        try await client.invoke("repository_is_pex_enabled", with: handle).boolValue.orThrow
    } }
    func setPex(enabled: Bool) async throws {
        try await client.invoke("repository_set_pex_enabled", with: ["repository": handle,
                                                                     "enabled": .bool(enabled)])
    }

    /// Create a share token providing access to this repository with the given mode.
    func share(for mode: AccessMode, using secret: Secret? = nil) async throws -> ShareToken {
        try await ShareToken(client, client.invoke("repository_share",
                                                   with: ["repository": handle,
                                                          "secret": secret?.value ?? .nil,
                                                          "mode": .uint(UInt64(mode.rawValue))]))
    }

    var syncProgress: Progress { get async throws {
        try await Progress(client.invoke("repository_sync_progress", with: handle))
    } }

    var infoHash: String { get async throws {
        try await client.invoke("repository_get_info_hash", with: handle).stringValue.orThrow
    } }

    /// Create mirror of this repository on a cache server.
    func createMirror(to host: String) async throws {
        try await client.invoke("repository_create_mirror", with: ["repository": handle,
                                                                   "host": .string(host)])
    }

    /// Check if this repository is mirrored on a cache server.
    func mirrorExists(on host: String) async throws -> Bool {
        try await client.invoke("repository_mirror_exists",
                                with: ["repository": handle,
                                       "host": .string(host)]).boolValue.orThrow
    }

    /// Delete mirror of this repository from a cache server.
    func deleteMirror(from host: String) async throws {
        try await client.invoke("repository_delete_mirror", with: ["repository": handle,
                                                                   "host": .string(host)])
    }

    func metadata(for key: String) async throws -> String? {
        let res = try await client.invoke("repository_get_metadata", with: ["repository": handle,
                                                                            "key": .string(key)])
        if case .nil = res { return nil }
        return try res.stringValue.orThrow
    }

    /// Performs an (presumably atomic) CAS on `edits`, returning `true` if they were updated
    func updateMetadata(with edits: [String:(from: String, to: String)]) async throws -> Bool {
        try await client.invoke("repository_set_metadata",
                                with: ["repository": handle,
                                       "edits": .array(edits.map {["key": .string($0.key),
                                                                   "old": .string($0.value.from),
                                                                   "new": .string($0.value.to)]})]
        ).boolValue.orThrow
    }

    /// Mount the repository if supported by the platform.
    @available(*, deprecated, message: "Not supported on darwin")
    func mount() async throws {
        try await client.invoke("repository_mount", with: handle)
    }

    /// Unmount the repository.
    @available(*, deprecated, message: "Not supported on darwin")
    func unmount() async throws {
        try await client.invoke("repository_unmount", with: handle)
    }

    /// The mount point of this repository (or `nil` if not mounted).
    @available(*, deprecated, message: "Not supported on darwin")
    var mountPoint: String? { get async throws {
        let res = try await client.invoke("repository_get_mount_point", with: handle)
        if case .nil = res { return nil }
        return try res.stringValue.orThrow
    } }

    var networkStats: NetworkStats { get async throws {
        try await NetworkStats(client.invoke("repository_get_stats", with: handle))
    } }

    // FIXME: nominally called setAccess which is easily confused with setAccessMode
    /** Updates the secrets used to access this repository
     *
     * The default value maintains the existing key whereas explicitly passing `nil` removes it.
     *
     * `Known issue`: keeping an existing `read password` while removing the `write password` can
     * result in a `read-only` repository. If you break it, you get to keep all the pieces!
     */
    func unsafeSetSecrets(readSecret: Secret? = KEEP_EXISTING,
                          writeSecret: Secret? = KEEP_EXISTING) async throws {
        // FIXME: the implementation requires a distinction between "no key" and "remove key"...
        /** ...which is currently implemented in terms of a `well known` default value
         *
         * On a related matter, we are currently leaking an _important_ bit in the logs (the
         * existence or lack thereof of some secret) because we can't override `Optional.toString`.
         *
         * While the root cause is different, many languages have a similar
         * `nil is a singleton` problem which we can however fix via convention:
         *
         * If we agree that a secret of length `0` means `no password`, we can then use `nil` here
         * as a default argument for `keep existing secret` on both sides of the ffi */
        func encode(_ arg: Secret?) -> MessagePackValue {
            guard let arg else { return .string("disable") }
            return arg.value == KEEP_EXISTING.value ? .nil : ["enable": readSecret!.value]
        }
        try await client.invoke("repository_set_access", with: ["repository": handle,
                                                                "read": encode(readSecret),
                                                                "write": encode(writeSecret)])
    }
}

// feel free to swap this with your preferred subset of the output of `head /dev/random | base64`
@usableFromInline let KEEP_EXISTING = Password("i2jchEQApyAkAD79uPYvuO1jiTAumhAwwSOQx5GGuNu3NZPKc")
